### PR TITLE
New resource: Grafana Workspace Api Key

### LINKED ISSUE #26374: Add resource to create an API key for managed grafana
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

I'm trying to create a managed Grafana instance using the AWS provider and then manage the Grafana dashboards/etc using the Grafana provider.

In order for the Grafana provider to authenticate with the Grafana instance, it needs to have an API key. As part of my infrastructure configuration, I'd like to ensure my Grafana provider can have access to an API.

The goal would be to use `aws_grafana_workspace_api_key` to create an admin key, store it in Secrets Manager, and then read that value as part of my Dashboard terraform configuration.

### New or Affected Resource(s)

* aws_grafana_workspace_api_key
* aws_grafana_workspace (needs to export the `id` for use with the new resource)

### Potential Terraform Configuration

```hcl
resource "aws_grafana_workspace" "example" {
  account_access_type      = "CURRENT_ACCOUNT"
  authentication_providers = ["SAML"]
  permission_type          = "SERVICE_MANAGED"
  role_arn                 = aws_iam_role.assume.arn
}

resource "aws_grafana_workspace_api_key" "terraform" {
  workspace_id = aws_grafana_workspace.example.id
  name            = "${var.environment}-terraform-api-key"
  role            = "Admin"
  seconds_to_live = 2592000 ### 1 month
}

resource "aws_secretsmanager_secret" "grafana-key" {
  name                    = "${var.environment}-grafana-key"
  recovery_window_in_days = 0
}

resource "aws_secretsmanager_secret_version" "grafana-key" {
  secret_id     = aws_secretsmanager_secret.grafana-key.id
  secret_string = jsonencode({
    key   = "${var.environment}-grafana-secret"
    value = aws_grafana_workspace_api_key.terraform
  })
}
```

### References

* Create a workspace key: https://docs.aws.amazon.com/grafana/latest/APIReference/API_CreateWorkspaceApiKey.html)
* Delete a workspace key: https://docs.aws.amazon.com/grafana/latest/APIReference/API_DeleteWorkspaceApiKey.html

### PR DESCRIPTION
**Pull request by bschaatsbergen**
### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for pull request followers and do not help prioritize the request

Closes #26374

Output from acceptance testing:

```
bruno $ make testacc TESTS=TestAccWorkspaceApiKey_basic PKG=grafana
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./internal/service/grafana/... -v -count 1 -parallel 20 -run='TestAccWorkspaceApiKey_basic'  -timeout 180m
=== RUN   TestAccWorkspaceApiKey_basic
--- PASS: TestAccWorkspaceApiKey_basic (147.87s)
PASS
ok      github.com/hashicorp/terraform-provider-aws/internal/service/grafana    147.980s
```

### CODE REVIEW

File: internal/service/grafana/workspace_api_key_test.go [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+	"fmt"
+	"testing"
+
+	"github.com/aws/aws-sdk-go/service/managedgrafana"
+	sdkacctest "github.com/hashicorp/terraform-plugin-sdk/v2/helper/acctest"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/resource"
+	"github.com/hashicorp/terraform-provider-aws/internal/acctest"
+)
+
+func testAccWorkspaceApiKeyConfig_basic(t *testing.T) {
```
Comment (ewbankkit) [inline internal/service/grafana/workspace_api_key_test.go:13]:
```suggestion
func testAccWorkspaceApiKey_basic(t *testing.T) {
```

File: internal/service/grafana/workspace_api_key_test.go [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+					resource.TestCheckResourceAttr(resourceName, "key_name", rName),
+					resource.TestCheckResourceAttr(resourceName, "key_role", "EDITOR"),
+					resource.TestCheckResourceAttr(resourceName, "seconds_to_live", "3600"),
+				),
+			},
+		},
+	})
+}
+
+func testAccWorkspaceApiKey_providerBasic(rName string) string {
```
Comment (ewbankkit) [inline internal/service/grafana/workspace_api_key_test.go:37]:
```suggestion
func testAccWorkspaceApiKeyConfig_providerBasic(rName string) string {
```

File: internal/service/grafana/workspace_api_key_test.go [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+	workspaceResourceName := "aws_grafana_workspace.test"
+
+	resource.Test(t, resource.TestCase{
+		PreCheck:                 func() { acctest.PreCheck(t); acctest.PreCheckPartitionHasService(managedgrafana.EndpointsID, t) },
+		ErrorCheck:               acctest.ErrorCheck(t, managedgrafana.EndpointsID),
+		CheckDestroy:             nil,
+		ProtoV5ProviderFactories: acctest.ProtoV5ProviderFactories,
+		Steps: []resource.TestStep{
+			{
+				Config: testAccWorkspaceApiKey_providerBasic(rName),
```
Comment (ewbankkit) [inline internal/service/grafana/workspace_api_key_test.go:25]:
```suggestion
				Config: testAccWorkspaceApiKeyConfig_providerBasic(rName),
```

File: website/docs/r/grafana_workspace_api_key.html.markdown [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+The following arguments are required:
+
+- `workspace_id` - (Required) The ID of the workspace that the key is valid for.
+- `key_name` - (Required) Specifies the name of the key. Keynames must be unique to the workspace.
+- `key_role` - (Required) Specifies the permission level of the key. Valid values are `VIEWER`, `EDITOR`, or `ADMIN`.
+- `seconds_to_live` - (Required) Specifies the time in seconds until the key expires. Keys can be valid for up to 30 days.
+
+## Attributes Reference
+
+All the above arguments are also exported as attributes.
```
Comment (allenhumphreys) [inline website/docs/r/grafana_workspace_api_key.html.markdown:37]: Should be:

No additional attributes are exported.

File: internal/service/grafana/workspace_api_key.go [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+	d.Set("seconds_to_live", d.Get("seconds_to_live").(int))
+	d.Set("workspace_id", d.Get("workspace_id").(string))
+
+	return nil
+}
+
+func resourceWorkspaceApiKeyDelete(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*conns.AWSClient).GrafanaConn
+
+	log.Printf("[DEBUG] Deleting Grafana Workspace API Key: %s", d.Id())
```
Comment (allenhumphreys) [inline internal/service/grafana/workspace_api_key.go:84]: API -> Api

File: internal/service/grafana/workspace_api_key.go [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+	_, err := conn.DeleteWorkspaceApiKey(&managedgrafana.DeleteWorkspaceApiKeyInput{
+		KeyName:     aws.String(d.Get("key_name").(string)),
+		WorkspaceId: aws.String(d.Get("workspace_id").(string)),
+	})
+
+	if err != nil {
+		return fmt.Errorf("error deleting Grafana Workspace API Key (%s): %w", d.Id(), err)
```
Comment (allenhumphreys) [inline internal/service/grafana/workspace_api_key.go:91]: API -> Api

**Comment by allenhumphreys** [comment]
@bschaatsbergen Sorry for misinterpreting the inscrutable CI output:

Running this locally helped `semgrep --max-lines-per-finding 1 --config .ci/.semgrep-caps-aws-ec2.yml`

```
Findings:

  internal/service/grafana/workspace_api_key.go 
     ci.caps2-in-func-name
        Use correct caps in func name (i.e., HTTPS or https, not Https) (see list at
        https://github.com/hashicorp/terraform-provider-aws/tree/main/names/caps.md)

         14┆ func ResourceWorkspaceApiKey() *schema.Resource {
         50┆ func resourceWorkspaceApiKeyCreate(d *schema.ResourceData, meta interface{}) error {
         72┆ func resourceWorkspaceApiKeyRead(d *schema.ResourceData, meta interface{}) error {
         81┆ func resourceWorkspaceApiKeyDelete(d *schema.ResourceData, meta interface{}) error {


  internal/service/grafana/workspace_api_key_test.go 
     ci.caps2-in-func-name
        Use correct caps in func name (i.e., HTTPS or https, not Https) (see list at
        https://github.com/hashicorp/terraform-provider-aws/tree/main/names/caps.md)

         13┆ func testAccWorkspaceApiKey_basic(t *testing.T) {
         37┆ func testAccWorkspaceApiKeyConfig_providerBasic(rName string) string {
```

You can see these 6 function names need to actually be updated from `Api` to `API` to pass the checks!

**Comment by bschaatsbergen** [comment]
> @bschaatsbergen Sorry for misinterpreting the inscrutable CI output!
> 
> Running this locally helped `semgrep --max-lines-per-finding 1 --config .ci/.semgrep-caps-aws-ec2.yml`
> 
> ```
> Findings:
> 
>   internal/service/grafana/workspace_api_key.go 
>      ci.caps2-in-func-name
>         Use correct caps in func name (i.e., HTTPS or https, not Https) (see list at
>         https://github.com/hashicorp/terraform-provider-aws/tree/main/names/caps.md)
> 
>          14┆ func ResourceWorkspaceApiKey() *schema.Resource {
>          50┆ func resourceWorkspaceApiKeyCreate(d *schema.ResourceData, meta interface{}) error {
>          72┆ func resourceWorkspaceApiKeyRead(d *schema.ResourceData, meta interface{}) error {
>          81┆ func resourceWorkspaceApiKeyDelete(d *schema.ResourceData, meta interface{}) error {
> 
> 
>   internal/service/grafana/workspace_api_key_test.go 
>      ci.caps2-in-func-name
>         Use correct caps in func name (i.e., HTTPS or https, not Https) (see list at
>         https://github.com/hashicorp/terraform-provider-aws/tree/main/names/caps.md)
> 
>          13┆ func testAccWorkspaceApiKey_basic(t *testing.T) {
>          37┆ func testAccWorkspaceApiKeyConfig_providerBasic(rName string) string {
> ```
> 
> You can see these 6 function names need to actually be updated from `Api` to `API` to pass the checks!

Fixed

**Comment by ewbankkit** [comment]
@bschaatsbergen Thanks for the contribution :tada: :clap:.
I have incorporated your work into https://github.com/hashicorp/terraform-provider-aws/pull/25286.

**Comment by bschaatsbergen** [comment]
Thanks, that's fine! @ewbankkit

**Pull request closed without merging**